File: maxPoints.cpp


// leetcode.com 149. Max Points on a Line

#include <vector>
#include <map>


struct Point {
     int x;
     int y;
     Point() : x(0), y(0) {}
     Point(int a, int b) : x(a), y(b) {}
};

using namespace std ;

class Solution {
public:
    int maxPoints(vector<Point>& points) {
    	int n = points.size() ;
    	int ans = 0 ;

    	for(int i = 0 ; i < n ; ++i)
    	{
    		map<pair<int, int>, int> count ;
    		int same_points = 1 ;
    		int max_points = 0 ;

    		for(int j = i + 1 ; j < n ; ++j)
    		{
    			const Point& p1 = points[i] ;
    			const Point& p2 = points[j] ;

    			if(p1.x == p2.x && p1.y == p2.y) 
    				++same_points ;
    			else
    				max_points = max(max_points, ++count[getSlope(p1, p2)]) ;
    		}

    		ans = max(ans, same_points + max_points) ;
    	}

    	return ans ;	    
    }

private:
	pair<int, int> getSlope(const Point& p1, const Point& p2)
	{
		const int dx = p2.x - p1.x ;
		const int dy = p2.y - p1.y ;

		if(dy == 0) return {p1.y, 0} ;

		if(dx == 0) return {0, p1.x} ;

		int d = gcd(dx, dy) ;

		return {dy / d, dx / d} ;
	}

	int gcd(int m, int n)
	{
		return n == 0 ? m : gcd(n, m % n) ;
	}
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p100/MaxPoints.java
package com.vkeonline.leetcode.p100;

import java.util.HashMap;
import java.util.Map;

/**
 * Leetcode: 149. Max Points on a Line
 * @author csgear
 */
public class MaxPoints {
    public int maxPoints(int[][] points) {
        if (points == null || points.length == 0) {
            return 0;
        }
        Map<Double, Integer> map = new HashMap<>();
        int max = 1;
        for (int i = 0; i < points.length; i++) {
            map.clear();
            int dup = 0;
            map.put(Double.MIN_VALUE, 1);
            for (int j = i + 1; j < points.length; j++) {
                if (points[i][0] == points[j][0] && points[i][1] == points[j][1]) {
                    dup++;
                    continue;
                }
                double k = points[i][0] - points[j][0] == 0 ?
                        Integer.MAX_VALUE : 0.0 + (double) (points[i][1] - points[j][1]) / (points[i][0] - points[j][0]);
                map.put(k, map.getOrDefault(k, 1) + 1);
            }
            for (Integer count : map.values()) {
                if (count + dup > max) {
                    max = count + dup;
                }
            }
        }
        return max;
    }
}
